🎉 Initial Setup

burgholzer opened the pull request:
This is the initial draft of a setup for this kind of action.
A lot of experimenting is probably going to be done here.

denialhaag commented on README.md:
Once everything is finalized, I will update the README.

burgholzer commented on .github/workflows/build-portable-mlir-toolchain.yml:
```diff
@@ -0,0 +1,58 @@
+
+  release:
+    needs: [linux, macos, windows]
+    if: ${{ inputs.release == true || true }}
```
This sounds reasonable! 👍🏼

burgholzer commented on .github/workflows/build-portable-mlir-toolchain.yml:
```diff
@@ -0,0 +1,58 @@
+          gh release create ${{ inputs.tag || 'test-release-4'}} \
+                            artifacts/**/* \
+                            --title "Test Release" \
+                            --notes "This is a test release. It will be deleted."
```
The title could probably be automates similarly to the release drafter, right? It would be the current date in CalVer format as a tag and as part of the release title.
Probably, the notes could be autogenerated as well.

denialhaag replied:
I have added a proposal for an auto-generated release title now.

I'm not so sure about the release notes. Ideally, we want to specify quite a few things there. This would be the only way to inform about changes to the build options, the used LLVM ref (which, to be fair, could be taken from the inputs), and potential changes to the GitHub Action. 

What is your take on this? 🤔

burgholzer commented on .license-tools-config.json:
```diff
@@ -0,0 +1,40 @@
+{
+  "author": {
+    "name": "Lukas Burgholzer",
+    "years": [2025]
```
We should update this before finalising.
Either your name, @denialhaag, should appear here too, or we go with MQSC or MQSC+TUM.

burgholzer commented on LICENSE.md:
Do we need to use the version with the LLVM exceptions that LLVM itself uses? 
Probably makes sense. Inspiration for that can be found here:
- https://github.com/Munich-Quantum-Software-Stack/QDMI/blob/develop/LICENSE

> SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

denialhaag replied:
Done!

burgholzer commented on package.json:
```diff
@@ -0,0 +1,33 @@
+  },
+  "keywords": [],
+  "author": "burgholzer",
+  "license": "ISC",
```
We definitely need to update this before finalising.
A couple other things in here need to be adjusted too once we move the repository elsewhere.

burgholzer commented on package.json:
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "portable-mlir-toolchain",
```
do we want to call this `setup-mlir`?
would be kind of fitting and idiomatic. 
Could also be a good name for the entire repository/project.

burgholzer commented on scripts/toolchain/windows/build.ps1:
```diff
@@ -0,0 +1,114 @@
+        '-DLLVM_INCLUDE_EXAMPLES=OFF',
+        '-DLLVM_ENABLE_ASSERTIONS=ON',
+        '-DLLVM_INSTALL_UTILS=ON',
+        "-DCMAKE_INSTALL_PREFIX=$install_prefix"
```
Same questions about LTO and the compiler being used here.
Additionally, this invocation does not enable RTTI as opposed to the others.

burgholzer commented on src/get-download-link.ts:
```diff
@@ -0,0 +1,123 @@
+        tag: tag
+    })
+    return response.data.assets
+}
```
just to make sure: `tag` here is the CalVer tag of the release in this repository here, right? and not the LLVM tag.
We should try to clarify this across the repository here to always make sure that it is clear, which version to specify where.

denialhaag replied:
Exactly! I will make sure to streamline this tomorrow! I would propose using `llvmorg_ref` and `llvmorg_tag` for all LLVM refs and tags. I haven't found a good prefix for our refs and tags yet. 🤔

denialhaag replied:
In principle, it would of course be nice if our action also worked with LLVM tags, but then we would need to store some kind of map from LLVM tags to our tags. Plus, if we ever have two releases for the same LLVM release, this could get a bit messy.

denialhaag replied:
I have updated all the variable names in https://github.com/munich-quantum-software/setup-mlir/pull/1/commits/05554d60fa3b37c9afc7a3a8f4279a1439eebe7b. 😌

burgholzer reviewed (commented):
Hi @denialhaag 👋🏼 
Thanks for all your work on this. Really great to see this come together 🙌🏼 
I know it isn't fully ready, but I still wanted to give this a quick review and leave some comments here. Hopefully, none of them take long to incorporate.

denialhaag commented on .license-tools-config.json:
```diff
@@ -0,0 +1,40 @@
+    "years": [2025]
+  },
+  "force_author": true,
+  "license": "MIT",
```
I have now updated it to a custom `Apache-2.0 WITH LLVM-exception` license. 👍

denialhaag commented on scripts/toolchain/linux/in-container.sh:
```diff
@@ -0,0 +1,112 @@
+    -DLLVM_BUILD_EXAMPLES=OFF \
+    -DLLVM_BUILD_TESTS=OFF \
+    -DLLVM_ENABLE_ASSERTIONS=ON \
+    -DLLVM_ENABLE_LTO=Thin \
```
```suggestion
    -DCMAKE_INSTALL_PREFIX="$INSTALL_PREFIX" \
    -DLLVM_BUILD_EXAMPLES=OFF \
    -DLLVM_BUILD_TESTS=OFF \
    -DLLVM_ENABLE_ASSERTIONS=ON \
```

denialhaag commented on .github/workflows/build-portable-mlir-toolchain.yml:
```diff
@@ -0,0 +1,58 @@
+        env:
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          gh release create ${{ inputs.tag || 'test-release-6'}} \
```
Just as a reminder, before it makes the next run fail. 😌 

```suggestion
          gh release create ${{ inputs.tag || 'test-release-7'}} \
```

denialhaag commented on scripts/toolchain/linux/Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+# Install build dependencies
+RUN yum -y update && \
+    yum -y install git cmake ninja-build python3 python3-pip curl zstd xz tar which gcc-c++ make perf ccache && \
```
This should also be done now.

denialhaag commented on scripts/toolchain/macos/build.sh:
```diff
@@ -0,0 +1,110 @@
+    -DLLVM_INSTALL_UTILS=ON \
+    -DLLVM_ENABLE_RTTI=ON \
+    -DCMAKE_INSTALL_PREFIX="$prefix" \
+    -DCMAKE_OSX_DEPLOYMENT_TARGET=${MACOSX_DEPLOYMENT_TARGET:-11.0}
```
The compiler is specified, and ThinLTO is enabled on ARM systems. On x86 systems, the build exceeded the 6-hour limit for GitHub Workflows.

burgholzer commented:
Moved the repository, renamed it, and put the PR out of draft.
A code rabbit review should be coming soon.
I'll try to get my own review in soon after.

denialhaag commented on src/get-download-link.ts:
```diff
@@ -0,0 +1,133 @@
+        tag: tag
+    })
+    return response.data.assets
+}
```
Yes, that's related to a previous comment. Done.

denialhaag commented on src/get-download-link.ts:
```diff
@@ -0,0 +1,133 @@
+    return response.data.assets
+}
+
+/**s
```
Done.

denialhaag commented on scripts/toolchain/macos/build.sh:
```diff
@@ -0,0 +1,141 @@
+  echo "Error: Installation directory (-p) is required"
+  echo "Usage: $0 -r <ref> -p <installation directory>"
+  exit 1
+fi
```
Done.

denialhaag commented on installation/setup-mlir.sh:
```diff
@@ -0,0 +1,120 @@
+  echo "Error: Installation directory (-p) is required"
+  echo "Usage: $0 -t <tag> -p <installation directory>"
+  exit 1
+fi
```
Done.

denialhaag commented on scripts/toolchain/linux/build.sh:
```diff
@@ -0,0 +1,88 @@
+# If ROOT_DIR is mounted at /work, then SCRIPT_DIR becomes /work${REL_DIR}
+REL_DIR="${SCRIPT_DIR#${ROOT_DIR}}"
+IN_CONTAINER_SCRIPT="/work${REL_DIR}/in-container.sh"
+
```
Done.

denialhaag commented on installation/setup-mlir.ps1:
```diff
@@ -0,0 +1,92 @@
+$ErrorActionPreference = "Stop"
+
+# Change to installation directory
+pushd $install_prefix > $null
```
Done, at least kind of. The script now fails if the directory doesn't exist.

denialhaag commented on installation/setup-mlir.sh:
```diff
@@ -0,0 +1,120 @@
+# Unpack archive
+echo "Extracting archive..."
+zstd -d "asset.tar.zst" --output-dir-flat .
+tar -xf "asset.tar"
```
Done.

denialhaag commented on package.json:
```diff
@@ -0,0 +1,33 @@
+  "main": "./dist/index.js",
+  "scripts": {
+    "build": "npx ncc build src/index.ts -o dist --source-map --license licenses.txt",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
I have removed the test script altogether now.

denialhaag commented on src/index.ts:
```diff
@@ -0,0 +1,60 @@
+  } else if (error instanceof Error) {
+    core.setFailed(error.message)
+  }
+}
```
Done.

denialhaag commented on action.yml:
```diff
@@ -0,0 +1,41 @@
+    default: ${{ github.token }}
+runs:
+  using: "node24"
+  main: "dist/index.js"
```
Done.

denialhaag commented on src/get-download-link.ts:
```diff
@@ -0,0 +1,133 @@
+  } else {
+    throw new Error(`Unsupported architecture: ${process.arch}`)
+  }
+}
```
Done.

denialhaag commented on scripts/toolchain/windows/build.ps1:
```diff
@@ -0,0 +1,162 @@
+
+# Extract archive
+Write-Host "Extracting archive into $repo_dir..."
+tar -xzf $temp_archive --strip-components=1 -C $repo_dir
```
Done in my last commit.

denialhaag commented on scripts/toolchain/windows/build.ps1:
```diff
@@ -0,0 +1,162 @@
+        'lld*.exe',
+        'llvm-bolt.exe',
+        'perf2bolt.exe'
+    )
```
Done.

denialhaag commented on scripts/toolchain/linux/build.sh:
```diff
@@ -0,0 +1,88 @@
+    exit 1
+    ;;
+  esac
+done
```
I have removed this logic now. The arguments are checked later as well.

denialhaag commented on scripts/toolchain/macos/build.sh:
```diff
@@ -0,0 +1,138 @@
+    p) INSTALL_PREFIX="$OPTARG"
+    ;;
+  esac
+done
```
Done.

denialhaag commented on installation/setup-mlir.sh:
```diff
@@ -0,0 +1,129 @@
+    a) GITHUB_TOKEN="$OPTARG"
+    ;;
+  esac
+done
```
Done.

denialhaag commented on src/get-download-link.ts:
```diff
@@ -0,0 +1,133 @@
+      repo: "setup-mlir",
+      tag: tag
+    })
+    return response.data.assets
```
Ahh, the files are no longer necessary. Removed.

burgholzer commented on installation/setup-mlir.sh:
```diff
@@ -0,0 +1,131 @@
+                   -H "Accept: application/vnd.github+json" \
+                   ${GITHUB_TOKEN:+-H "Authorization: Bearer $GITHUB_TOKEN"} \
+                   -H "X-GitHub-Api-Version: 2022-11-28" \
+                   "$ASSETS_URL")
```
In the Windows script, the token is truly optional. Here, it is set unconditionally. Is that on purpose?

denialhaag replied:
If I understood correctly, the token is only added to headers if `GITHUB_TOKEN` is set. Otherwise, the entire line resolves to null, and an unauthenticated request is made. 🤔

burgholzer commented on package.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "setup-mlir",
+  "version": "1.0.0",
+  "description": "Portable MLIR toolchain",
```
This should probably be updated.

burgholzer reviewed (commented):
Many thanks @denialhaag for all the work on this.
A final round of comments, before this can go in and we can release a first version. Should just be minor stuff.

denialhaag commented on installation/setup-mlir.ps1:
```diff
@@ -0,0 +1,102 @@
+if (-not (Get-Command tar -ErrorAction SilentlyContinue)) {
+    Write-Error "tar not found. Please install tar (e.g., via Chocolatey: choco install tar)."
+    exit 1
+}
```
Done.

denialhaag commented on installation/setup-mlir.ps1:
```diff
@@ -0,0 +1,102 @@
+
+# Clean up
+Remove-Item "asset.tar.zst" -Force
+Remove-Item "asset.tar" -Force
```
Done.

burgholzer commented on .github/workflows/build-portable-mlir-toolchain.yml:
```diff
@@ -0,0 +1,63 @@
+                            artifacts/**/* \
+                            installation/* \
+                            --title "Setup MLIR ${{ inputs.release-tag }} Release" \
+                            --notes "${{ inputs.release-notes }}"
```
`installation/` is a local directory that comes from the repo checkout and contains the installation scripts.

burgholzer commented on .github/workflows/build-portable-mlir-toolchain.yml:
```diff
@@ -0,0 +1,63 @@
+      release-notes:
+        description: "Release notes for the GitHub release"
+        required: true
+        type: string
```
The respective job step (release) only runs via the `workflow_dispatch` _and_ when the `release` input is `true`.
So it should be safe to require the other inputs.
